src/builtindict.js

```diff
@@ -229,19 +229,40 @@ Sk.abstr.setUpModuleMethods("builtins", Sk.builtins, {
 
     eval_$rw$: {
         $name: "eval",
-        $meth: Sk.builtin.eval_,
+        $meth: function (source, globals, locals) {
+            // check globals
+            const tmp_globals = checkGlobLoc(globals, "globals");
+            // check locals
+            const tmp_locals = checkGlobLoc(locals, "locals");
+            return Sk.misceval.chain(Sk.builtin.eval(source, tmp_globals, tmp_locals), (res) => {
+                reassignGlobLoc(globals, tmp_globals);
+                reassignGlobLoc(locals, tmp_locals);
+                return res;
+            });
+        },
         $flags: { MinArgs: 1, MaxArgs: 3 },
         $textsig: "($module, source, globals=None, locals=None, /)",
         $doc:
             "Evaluate the given source in the context of globals and locals.\n\nThe source may be a string representing a Python expression\nor a code object as returned by compile().\nThe globals must be a dictionary and locals can be any mapping,\ndefaulting to the current globals and locals.\nIf only globals is given, locals defaults to it.",
     },
 
-    // exec: {
-    //     $meth: Sk.builtin.exec,
-    //     $flags: {MinArgs:2, MaxArgs: 3},
-    //     $textsig: "($module, source, globals=None, locals=None, /)",
-    //     $doc: "Execute the given source in the context of globals and locals.\n\nThe source may be a string representing one or more Python statements\nor a code object as returned by compile().\nThe globals must be a dictionary and locals can be any mapping,\ndefaulting to the current globals and locals.\nIf only globals is given, locals defaults to it."
-    // },
+    exec: {
+        $meth: function (source, globals, locals) {
+            // check globals
+            const tmp_globals = checkGlobLoc(globals, "globals");
```

Comment:
`checkGlobLoc` is a slightly weird name for this function, which actually converts from py dict to JS obj/scope